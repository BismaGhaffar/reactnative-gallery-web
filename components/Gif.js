import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

export const Smartphone = styled.div`
  position: relative;
  padding: 65px 12px;
  background: #404040;
  border-radius: 38px;
  box-shadow: inset 0 0 3px 0 rgb(79, 86, 95, 0.2);
  border: solid 1px rgb(79, 86, 95, 0.2);
  max-width: 275px;
  min-width: 170px;
  min-height: 250px;
  cursor: ${({ cursorPointer }) =>
    (cursorPointer ? 'pointer' : 'url(/static/images/play-circle.svg), auto;')};
`

export const GifContainer = styled.div`
  position: relative;
  padding: 0px;
  font-size: 0px;
  max-width: 250px;
  margin: 0px auto;
`

export const Sizer = styled.div`
  position: relative;
  padding-bottom: 180%;
`

const Play = styled.div`
  @media (max-width: 769px) {
    position: absolute;
    z-index: 100;
    margin: 0px auto;
    width: 100%;
    height: 100%;
    max-width: 250px;
    opacity: 0.5;
    content: url(/static/images/play-circle.svg);
    display: ${props => (props.show ? 'block' : 'none')};
  }
`

const ButtonContainer = styled.div`
  position: absolute;
  width: 100%;
  height: 45px;
  right: 0;
  padding-top: 7px;
`

const Button = styled.div`
  margin: 0px auto;
  border: solid 3px #707070;
  width: 45px;
  height: 45px;
  border-radius: 90px;
`

class Gif extends Component {
  static propTypes = {
    gifId: PropTypes.string.isRequired
  }

  state = {
    play: false,
    mouseover: false
  }

  onMouseEnterHandler = () => {
    this.video.play()
    this.setState({ play: true, mouseover: true })
  }

  onMouseLeaveHandler = () => {
    this.video.pause()
    this.setState({ play: false, mouseover: false })
  }

  onClick = () => {
    if (this.state.play && !this.state.mouseover) {
      this.video.pause()
      this.setState({ play: false })
    } else {
      this.video.play()
      this.setState({ play: true })
    }
  }

  render() {
    const { gifId } = this.props
    return (
      <Smartphone
        onMouseEnter={this.onMouseEnterHandler}
        onMouseLeave={this.onMouseLeaveHandler}
        onClick={this.onClick}
        style={{
          background: this.state.play && '#141414'
        }}
      >
        <GifContainer>
          <Play show={!this.state.play} />
          <Sizer />
          <video
            ref={(ref) => {
              this.video = ref
              return undefined
            }}
            autoPlay={false}
            loop
            playsInline
            muted
            poster={`https://thumbs.gfycat.com/${gifId}-poster.jpg`}
            style={{
              width: '100%',
              height: '100%',
              position: 'absolute',
              top: 0,
              left: 0
            }}
          >
            <track kind="captions" />
            <source
              src={`https://giant.gfycat.com/${gifId}.webm`}
              type="video/webm"
            />
            <source
              src={`https://giant.gfycat.com/${gifId}.mp4`}
              type="video/mp4"
            />
          </video>
        </GifContainer>
        <ButtonContainer>
          <Button />
        </ButtonContainer>
      </Smartphone>
    )
  }
}

export default Gif
