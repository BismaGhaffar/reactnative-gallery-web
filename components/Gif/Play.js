import styled from 'styled-components'

const Play = styled.div`
  @media (max-width: 769px) {
    position: absolute;
    z-index: 100;
    margin: 0px auto;
    width: 100%;
    height: 100%;
    max-width: 275px;
    opacity: 0.5;
    content: url(/static/images/play-circle.svg);
    display: ${props => (props.show ? 'block' : 'none')};
  }
`

export default Play
