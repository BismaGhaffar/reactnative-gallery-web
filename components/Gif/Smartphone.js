import styled from 'styled-components'

const Smartphone = styled.div`
  position: relative;
  padding: 65px 12px;
  background: rgba(49, 49, 49, 0.8);
  border-radius: 38px;
  box-shadow: inset 0 0 3px 0 rgb(79, 86, 95, 0.2);
  border: solid 1px rgb(79, 86, 95, 0.2);
  max-width: 275px;
  min-width: ${({ minWidth }) => (minWidth && `${minWidth}px`) || '194px'};
  min-height: 250px;
  margin: auto;
  cursor: ${({ cursorPointer }) =>
    (cursorPointer ? 'pointer' : 'url(/static/images/play-circle.svg), auto;')};
  transform: ${({ rotate }) => (rotate ? 'rotate(-0.25turn)' : 'none')};
  margin-top: ${({ rotate }) => (rotate ? '-100px' : 'auto')};
`

export default Smartphone
