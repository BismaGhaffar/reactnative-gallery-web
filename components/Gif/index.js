import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Smartphone from './Smartphone'
import GifContainer from './GifContainer'
import Play from './Play'
import { Button, ButtonContainer } from './Button'

export const Sizer = styled.div`
  position: relative;
  padding-bottom: 180%;
`

const Source = ({ mediatype, id }) => (
  <source
    src={`https://giant.gfycat.com/${id}.${mediatype}`}
    type={`video/${mediatype}`}
  />
)
Source.propTypes = {
  mediatype: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired
}

class Gif extends Component {
  static propTypes = {
    gifId: PropTypes.string.isRequired
  }

  state = {
    play: false,
    mouseover: false
  }

  onMouseEnterHandler = () => {
    this.play()
    this.setState({ play: true, mouseover: true })
  }

  onMouseLeaveHandler = () => {
    this.pause()
    this.setState({ play: false, mouseover: false })
  }

  onClick = () => {
    if (this.state.play && !this.state.mouseover) {
      this.pause()
      this.setState({ play: false })
    } else {
      this.play()
      this.setState({ play: true })
    }
  }

  pause = () => {
    this.video.pause()
  }

  play = () => {
    this.video.play()
  }

  render() {
    const { gifId } = this.props
    return (
      <Smartphone
        onMouseEnter={this.onMouseEnterHandler}
        onMouseLeave={this.onMouseLeaveHandler}
        onClick={this.onClick}
        style={{
          background: this.state.play && '#141414'
        }}
      >
        <GifContainer>
          <Play show={!this.state.play} />
          <Sizer />
          <video
            ref={(ref) => {
              this.video = ref
              return undefined
            }}
            autoPlay={false}
            loop
            playsInline
            preload="none"
            muted
            poster={`https://thumbs.gfycat.com/${gifId}-poster.jpg`}
            style={{
              width: '100%',
              height: '100%',
              position: 'absolute',
              top: 0,
              left: 0
            }}
          >
            <track kind="captions" />
            <Source id={gifId} mediatype="webm" />
            <Source id={gifId} mediatype="mp4" />
          </video>
        </GifContainer>
        <ButtonContainer>
          <Button />
        </ButtonContainer>
      </Smartphone>
    )
  }
}

export default Gif
