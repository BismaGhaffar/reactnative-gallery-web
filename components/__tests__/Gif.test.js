import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Gif from '../Gif'

Enzyme.configure({ adapter: new Adapter() })

it('Gif can be created', () => {
  const comp = renderer.create(<Gif
    gifId="test"
    username="xcarpentier"
    slug="slug"
  />)
  expect(comp).toBeDefined()
})

it('<Gif /> toMatchSnapshot', () => {
  const tree = renderer.create(<Gif
    gifId="test"
    username="xcarpentier"
    slug="slug"
  />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Gif /> toMatchSnapshot click', () => {
  const gif = shallow(<Gif gifId="test" username="xcarpentier" slug="slug" />)

  const inst = gif.instance()
  inst.video = {
    play: () => 'play',
    pause: () => 'pause'
  }

  // start on click
  gif.simulate('click')
  expect(gif.state('play')).toBe(true)

  // stop on click
  gif.simulate('click')
  expect(gif.state('play')).toBe(false)

  // mouseenter
  gif.simulate('mouseenter')
  expect(gif.state('play')).toBe(true)
  expect(gif.state('mouseover')).toBe(true)

  // mouseleave
  gif.simulate('mouseleave')
  expect(gif.state('play')).toBe(false)
  expect(gif.state('mouseover')).toBe(false)
})
