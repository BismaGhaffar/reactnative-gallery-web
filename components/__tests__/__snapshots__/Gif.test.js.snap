// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gif /> toMatchSnapshot 1`] = `
<div
  className="Smartphone-s165niw0-0 bMKjgM"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "background": false,
    }
  }
>
  <div
    className="GifContainer-s1wbubqp-0 fvwRPH"
  >
    <div
      className="Play-s10ikt1c-0 gLlmwW"
    />
    <div
      className="Gif__Sizer-s1j91rz1-0 gpJmvZ"
    />
    <video
      autoPlay={false}
      loop={true}
      muted={true}
      playsInline={true}
      poster="https://thumbs.gfycat.com/test-poster.jpg"
      preload="none"
      style={
        Object {
          "height": "100%",
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <track
        kind="captions"
      />
      <source
        src="https://giant.gfycat.com/test.webm"
        type="video/webm"
      />
      <source
        src="https://giant.gfycat.com/test.mp4"
        type="video/mp4"
      />
    </video>
  </div>
  <div
    className="Button__ButtonContainer-s1yo1pni-0 kZcYqm"
  >
    <div
      className="Button-s1yo1pni-1 ewehdc"
      onClick={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
    />
  </div>
</div>
`;
